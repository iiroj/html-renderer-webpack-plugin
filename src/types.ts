export type RendererArgs = Partial<{
  assets: Record<string, string[] | undefined>;
  compilationAssets: import("webpack").Compilation["assets"];
  filename: string;
  options: Record<string, any>;
  path: string;
  publicPath: string;
  stats: any;
}>;

export type Renderer = (args: RendererArgs) => string | Promise<string>;

export type Options = Partial<{
  options: any;
  paths: string[] | (() => Promise<string[]>);
  renderer: Renderer | string;
}>;
