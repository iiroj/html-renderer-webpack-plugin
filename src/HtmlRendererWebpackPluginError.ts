import { WebpackError } from "webpack";

class HtmlRendererWebpackPluginError extends WebpackError {
  public constructor(message: string) {
    super(`HtmlRendererWebpackPlugin Error:\n\n${message}`);
  }
}

export default HtmlRendererWebpackPluginError;
