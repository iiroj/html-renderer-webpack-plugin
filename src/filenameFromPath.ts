export default (path: string) =>
  path.replace(/\/$/, "/index").replace(/^\//, "") + ".html";
