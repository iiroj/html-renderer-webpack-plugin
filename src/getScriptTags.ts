export default (publichPath: string = "/", files: Array<string>) => {
  const prefix = publichPath.endsWith("/") ? publichPath : publichPath + "/";

  return files
    .map(
      (file: string) =>
        `<script defer type="text/javascript" src="${prefix}${file}"></script>`
    )
    .join("\n");
};
