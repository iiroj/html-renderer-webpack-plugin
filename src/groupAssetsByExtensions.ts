import { RawSource } from "webpack-sources";

interface ObjectOfRawsource {
  [key: string]: RawSource;
}
interface ObjectOfArrays {
  [key: string]: string[];
}

export default (assets: ObjectOfRawsource) =>
  Object.keys(assets).reduce((accumulator: ObjectOfArrays, asset: string) => {
    const ext = asset.slice(((asset.lastIndexOf(".") - 1) >>> 0) + 2);

    if (!accumulator[ext]) {
      accumulator[ext] = [];
    }

    accumulator[ext].push(asset);

    return accumulator;
  }, {});
