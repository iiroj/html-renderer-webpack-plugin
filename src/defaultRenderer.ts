import getScriptTags from "./getScriptTags";

export default async ({
  assets,
  publicPath
}: import("./HtmlRendererWebpackPlugin").RendererArgs) =>
  "<!doctype html>" +
  "<head>" +
  '<meta charset="utf-8">' +
  "<title>HtmlRendererWebpackPlugin</title>" +
  '<meta name="viewport" content="width=device-width, initial-scale=1">' +
  "</head>" +
  '<div id="root">' +
  "</div>" +
  getScriptTags(publicPath, assets ? assets.js : []);
