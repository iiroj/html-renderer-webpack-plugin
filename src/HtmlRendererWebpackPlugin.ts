import { FSWatcher } from "chokidar";
import { resolve } from "path";
import { Compiler, compilation } from "webpack";
import { RawSource } from "webpack-sources";

import defaultRenderer from "./defaultRenderer";
import filenameFromPath from "./filenameFromPath";
import groupAssetsByExtensions from "./groupAssetsByExtensions";
import HtmlRendererWebpackPluginError from "./HtmlRendererWebpackPluginError";
import { Renderer, Options, RendererArgs } from "./types";

const PLUGIN_NAME = "HtmlRendererWebpackPlugin";

const chokidarOptions = {
  ignoreInitial: true,
  disableGlobbing: true,
  persistent: true,
};

export default class HtmlRendererWebpackPlugin {
  private readonly options?: Record<string, any>;
  private readonly paths: string[];
  private readonly renderer: Renderer;
  private src?: string;

  public constructor({
    options,
    paths = ["/"],
    renderer = defaultRenderer,
  }: Options = {}) {
    this.options = options;
    this.paths = paths;
    this.renderer =
      typeof renderer === "function"
        ? renderer
        : this.createRequireAsyncRenderer(renderer);
  }

  private createRequireAsyncRenderer = (src: string) => {
    const resolved = resolve(process.cwd(), src);
    this.src = resolved;
    return (args: RendererArgs) => {
      const renderer = require(resolved);
      return renderer(args);
    };
  };

  private plugin = async (
    compilation: compilation.Compilation,
    done: Function
  ) => {
    const stats = compilation.getStats().toJson();
    const { publicPath } = compilation.outputOptions || "";
    const assets = groupAssetsByExtensions(compilation.assets);

    for (const path of this.paths) {
      const filename = filenameFromPath(path);

      try {
        const html = await this.renderer({
          assets,
          compilationAssets: compilation.assets,
          filename,
          options: this.options,
          path,
          publicPath,
          stats,
        });
        compilation.assets[filename] = new RawSource(html);
      } catch (error) {
        compilation.errors.push(
          new HtmlRendererWebpackPluginError(
            error.stack || error.message || error
          )
        );
      }
    }

    done();
  };

  public apply(compiler: Compiler) {
    if (this.src) {
      compiler.hooks.afterCompile.tap(PLUGIN_NAME, (compilation) => {
        compilation.fileDependencies.add(this.src!);
      });

      let watcher: FSWatcher;

      compiler.hooks.watchRun.tap(PLUGIN_NAME, () => {
        const chokidar = require("chokidar");
        watcher = chokidar.watch(this.src!, chokidarOptions);
        watcher.on("change", () => null); // trigger compilation
      });

      compiler.hooks.watchClose.tap(PLUGIN_NAME, () => {
        if (watcher) watcher.close();
      });
    }

    compiler.hooks.emit.tapAsync(PLUGIN_NAME, this.plugin);
  }
}
