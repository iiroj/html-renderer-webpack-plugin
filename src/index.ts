import HtmlRendererWebpackPlugin from "./HtmlRendererWebpackPlugin";

export { default as HtmlRendererWebpackPluginError } from "./HtmlRendererWebpackPluginError";

export { Options, Renderer, RendererArgs } from "./types";

export default HtmlRendererWebpackPlugin;
