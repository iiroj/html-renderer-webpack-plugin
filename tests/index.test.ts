/* eslint-disable @typescript-eslint/no-var-requires */

import HtmlRendererWebpackPlugin from "../src/HtmlRendererWebpackPlugin";
import { Options } from "../src/types";
import webpack from "webpack";
import MemoryFileSystem from "memory-fs";

const getWebpackConfig = (options: Options, config: Record<string, any> = {}) =>
  Object.assign(
    {
      entry: ["./tests/__mocks__/index.js"],
      mode: "development" as const,
      plugins: [new HtmlRendererWebpackPlugin(options)],
    },
    config
  );

const renderer = () => "test";

describe("HtmlRendererWebpackPlugin", () => {
  let fs: MemoryFileSystem;

  beforeEach(() => {
    fs = new MemoryFileSystem();
  });

  const compile = (options: ReturnType<typeof getWebpackConfig>) =>
    new Promise<webpack.Stats["compilation"]>((resolve, reject) => {
      const compiler = webpack(options);

      compiler.inputFileSystem = fs;
      compiler.outputFileSystem = fs;

      compiler.run((error, stats) => {
        if (error) reject(error);
        resolve(stats!.compilation);
      });
    });

  const readFile = (name: string) =>
    fs.readFileSync(`${process.cwd()}/dist/${name}`).toString();

  it("should render with default renderer", async () => {
    const paths = ["/", "/foo", "/bar/"];

    await compile(getWebpackConfig({ paths }));

    expect(readFile("index.html")).toMatchSnapshot();
    expect(readFile("foo.html")).toMatchSnapshot();
    expect(readFile("bar/index.html")).toMatchSnapshot();

    expect(() => readFile("bar.html")).toThrowError("ENOENT");
  });

  it("should render with custom renderer", async () => {
    const paths = ["/"];
    await compile(getWebpackConfig({ paths, renderer }));
    expect(readFile("index.html")).toMatchSnapshot();
  });

  it("should require renderer from string", async () => {
    const paths = ["/"];

    await compile(
      getWebpackConfig({ paths, renderer: "./tests/__mocks__/renderer.js" })
    );

    expect(readFile("index.html")).toMatchSnapshot();
  });

  it("should pass options to renderer", async () => {
    const paths = ["/"];

    await compile(
      getWebpackConfig({
        options: { foo: "bar" },
        paths,
        renderer: "./tests/__mocks__/options-renderer.js",
      })
    );

    expect(readFile("index.html")).toMatchSnapshot();
  });
});
