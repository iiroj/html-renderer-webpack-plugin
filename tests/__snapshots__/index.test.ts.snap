// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HtmlRendererWebpackPlugin should render with custom renderer 1`] = `"test"`;

exports[`HtmlRendererWebpackPlugin should render with default renderer 1`] = `
"<!doctype html>
<head>
  <meta charset=\\"utf-8\\">
  <title>HtmlRendererWebpackPlugin</title>
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<div id=\\"root\\"></div>
<script defer type=\\"text/javascript\\" src=\\"/main.js\\"></script>
<script defer type=\\"text/javascript\\" src=\\"/runtime~main.js\\"></script>"
`;

exports[`HtmlRendererWebpackPlugin should render with default renderer 2`] = `
"<!doctype html>
<head>
  <meta charset=\\"utf-8\\">
  <title>HtmlRendererWebpackPlugin</title>
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<div id=\\"root\\"></div>
<script defer type=\\"text/javascript\\" src=\\"/main.js\\"></script>
<script defer type=\\"text/javascript\\" src=\\"/runtime~main.js\\"></script>"
`;

exports[`HtmlRendererWebpackPlugin should render with default renderer 3`] = `
"<!doctype html>
<head>
  <meta charset=\\"utf-8\\">
  <title>HtmlRendererWebpackPlugin</title>
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
</head>
<div id=\\"root\\"></div>
<script defer type=\\"text/javascript\\" src=\\"/main.js\\"></script>
<script defer type=\\"text/javascript\\" src=\\"/runtime~main.js\\"></script>"
`;
